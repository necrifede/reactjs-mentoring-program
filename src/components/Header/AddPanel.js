import React, { useState } from 'react';
import { Button } from 'react-bootstrap';
import Logo from '../Logo';
import AddEditMovieModal from '../AddEditMovieModal';
import { Modal } from 'react-bootstrap';
import { PropTypes } from 'prop-types';

const AddPanel = () => {
    const [show, setShow] = useState(false);

    return (
        <div className="d-flex">
            <div className="me-auto p-2">
                <Logo />
            </div>
            <div>
                <Button variant="primary" onClick={() => setShow(true)}>
                    + Add Movie
                </Button>
                <AddEditMovieModal show={show} hideFunction={() => setShow(false)} />
            </div>
        </div>
    );
};

AddPanel.propTypes = {};

export default AddPanel;
