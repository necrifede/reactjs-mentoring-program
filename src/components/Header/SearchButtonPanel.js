import React from 'react';
import PropTypes from 'prop-types';
import Logo from '../Logo';
import { Button } from 'react-bootstrap';
import { Search } from 'react-bootstrap-icons';
import { useSearchParams } from 'react-router-dom';

const SearchButtonPanel = () => {
    const [, setSearchParams] = useSearchParams();

    return (
        <div className="d-flex">
            <div className="me-auto p-2">
                <Logo />
            </div>
            <div>
                <Button variant="primary" onClick={() => setSearchParams({ movie: '' })}>
                    <Search />
                </Button>
            </div>
        </div>
    );
};

SearchButtonPanel.propTypes = {
    unselectMovie: PropTypes.func,
};

export default SearchButtonPanel;
