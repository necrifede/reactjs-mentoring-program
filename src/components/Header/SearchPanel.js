import React, { useEffect, useState } from 'react';
import { Button, Col, Row } from 'react-bootstrap';
import { useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';

const SearchPanel = () => {
    const navigate = useNavigate();
    const search = useSelector((state) => state.criterias.search);
    const [word, setWord] = useState(search);

    useEffect(() => {
        setWord(search);
    }, [search]);

    return (
        <Row>
            <Col className="p-5">
                <form className="d-flex justify-content-center">
                    <label htmlFor="movieName">Find your movie</label>
                    <input id="movieName" type="text" value={word} onChange={(e) => setWord(e.target.value)} />

                    <Button type="submit" onClick={() => navigate(`${word}`)}>
                        Search
                    </Button>
                </form>
            </Col>
        </Row>
    );
};

SearchPanel.propTypes = {};

export default SearchPanel;
