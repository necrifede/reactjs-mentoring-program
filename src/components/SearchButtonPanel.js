import React from "react";
import PropTypes from "prop-types";
import Logo from "./Logo";
import { Button } from "react-bootstrap";
import { Search } from "react-bootstrap-icons";

const SearchButtonPanel = ({ unselectMovie = () => {} }) => {
    return (
        <div className="d-flex">
            <div className="me-auto p-2">
                <Logo />
            </div>
            <div>
                <Button variant="primary" onClick={() => unselectMovie()}>
                    <Search />
                </Button>
            </div>
        </div>
    );
};

SearchButtonPanel.propTypes = {
    unselectMovie: PropTypes.func,
};

export default SearchButtonPanel;
