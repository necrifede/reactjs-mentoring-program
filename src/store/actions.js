import { actionTypes as actionMovies } from './reducerMovies';
import { actionTypes as actionSelectedMovie } from './reducerSelectedMovie';
import { actionTypes as actionCriterias } from './reducerCriterias';

export const fetchMoviesAction = (payload) => ({ type: actionMovies.FETCH_MOVIES, payload });
export const setMoviesAction = (movies) => ({ type: actionMovies.SET_MOVIES, payload: movies });
export const createMovie = (movie) => ({ type: actionMovies.CREATE_MOVIE, payload: movie });
export const deleteMovie = (movie) => ({ type: actionMovies.DELETE_MOVIE, payload: movie });
export const updateMovie = (movie) => ({ type: actionMovies.UPDATE_MOVIE, payload: movie });

export const setSelectedMovie = (movie) => ({ type: actionSelectedMovie.SET_SELECTED_MOVIE, payload: movie });
export const clearSelectedMovie = () => ({ type: actionSelectedMovie.CLEAR_SELECTED_MOVIE });

export const setFilterGenres = (payload) => ({ type: actionCriterias.SET_CRITERIA_FILTER, payload });
export const setCriteriaSortBy = (payload) => ({ type: actionCriterias.SET_CRITERIA_SORT_BY, payload });
export const setCriteriaSortOrder = (payload) => ({ type: actionCriterias.SET_CRITERIA_SORT_ORDER, payload });
