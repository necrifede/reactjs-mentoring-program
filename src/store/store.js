import { applyMiddleware, combineReducers, createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { composeWithDevTools } from '@redux-devtools/extension';
import { moviesReducer } from './reducerMovies';
import { selectedMovieReducer } from './reducerSelectedMovie';
import { criteriasReducer } from './reducerCriterias';
import { rootSaga } from './sagas';

const rootReducer = combineReducers({
    movies: moviesReducer,
    selected: selectedMovieReducer,
    criterias: criteriasReducer,
});

const sagaMiddleware = createSagaMiddleware();

export const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(sagaMiddleware)));

sagaMiddleware.run(rootSaga);
