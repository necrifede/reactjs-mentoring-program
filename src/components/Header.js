import React, { useContext, useState } from "react";
import PropTypes from "prop-types";
import { Col, Row } from "react-bootstrap";
import AddPanel from "./AddPanel";
import SearchPanel from "./SearchPanel";
import SearchButtonPanel from "./SearchButtonPanel";
import MovieDetails from "./MovieDetails";
import { useSelectedMovie } from "../hooks/useSelectedMovie";

const Header = ({ addMovie = () => {} }) => {
    const [selectedMovie, setSelectedMovie] = useSelectedMovie();

    return (
        <Row>
            <Col>
                {selectedMovie ? (
                    <>
                        <SearchButtonPanel unselectMovie={() => setSelectedMovie(undefined)} />
                        <MovieDetails movie={selectedMovie} />
                    </>
                ) : (
                    <>
                        <AddPanel addMovie={addMovie} />
                        <SearchPanel />
                    </>
                )}
            </Col>
        </Row>
    );
};

Header.propTypes = {
    addMovie: PropTypes.func,
};

export default Header;
