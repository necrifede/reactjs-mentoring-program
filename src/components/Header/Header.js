import React, { useContext, useState } from 'react';
import PropTypes from 'prop-types';
import { Col, Row } from 'react-bootstrap';
import AddPanel from './AddPanel';
import SearchPanel from './SearchPanel';
import SearchButtonPanel from './SearchButtonPanel';
import MovieDetails from './MovieDetails';
import { useSelector } from 'react-redux';

const Header = ({ addMovie = () => {} }) => {
    const selectedMovie = useSelector((state) => state.selected);

    return (
        <Row>
            <Col>
                {selectedMovie ? (
                    <>
                        <SearchButtonPanel />
                        <MovieDetails movie={selectedMovie} />
                    </>
                ) : (
                    <>
                        <AddPanel addMovie={addMovie} />
                        <SearchPanel />
                    </>
                )}
            </Col>
        </Row>
    );
};

Header.propTypes = {
    addMovie: PropTypes.func,
};

export default Header;
