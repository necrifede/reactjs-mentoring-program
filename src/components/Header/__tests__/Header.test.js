import React from 'react';
import { act, fireEvent, render, screen } from '@testing-library/react';
import Header from '../Header';
import { createTestStore } from '../../../../settings/setupTests';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { clearSelectedMovie, setSelectedMovie } from '../../../store';
import SearchButtonPanel from '../SearchButtonPanel';

describe('<Header />', () => {
    let store;

    beforeEach(() => {
        store = createTestStore();
    });

    test('should Render default header', () => {
        const container = render(
            <Provider store={store}>
                <Header />
            </Provider>,
            { wrapper: BrowserRouter }
        );
        expect(container).toMatchSnapshot();
    });

    test('should Render header displaying a movie', () => {
        store.dispatch(setSelectedMovie({ id: 123, title: 'A movie', overview: 'An overview', genres: [] }));

        const container = render(
            <Provider store={store}>
                <Header />
            </Provider>,
            { wrapper: BrowserRouter }
        );

        expect(screen.getByText(/A movie/i)).toBeInTheDocument();
        expect(screen.getByText(/An overview/i)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    test('should Render header displaying a movie and unselecting it when state change', async () => {
        store.dispatch(setSelectedMovie({ id: 123, title: 'A movie', overview: 'An overview', genres: [] }));
        const FullHeader = () => (
            <Provider store={store}>
                <BrowserRouter>
                    <Header />
                </BrowserRouter>
            </Provider>
        );
        const container = render(<FullHeader />);

        expect(screen.queryByText(/A movie/i)).toBeInTheDocument();
        expect(screen.queryByText(/An overview/i)).toBeInTheDocument();

        await store.dispatch(clearSelectedMovie());

        expect(screen.queryByText(/A movie/i)).not.toBeInTheDocument();
        expect(screen.queryByText(/An overview/i)).not.toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
});
