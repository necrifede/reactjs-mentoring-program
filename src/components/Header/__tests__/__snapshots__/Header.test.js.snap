// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should Render default header 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="d-flex"
          >
            <div
              class="me-auto p-2"
            >
              <div>
                <b>
                  Netflix
                </b>
                roulette
              </div>
            </div>
            <div>
              <button
                class="btn btn-primary"
                type="button"
              >
                + Add Movie
              </button>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="p-5 col"
            >
              <form
                class="d-flex justify-content-center"
              >
                <label
                  for="movieName"
                >
                  Find your movie
                </label>
                <input
                  id="movieName"
                  type="text"
                  value=""
                />
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  Search
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="d-flex"
        >
          <div
            class="me-auto p-2"
          >
            <div>
              <b>
                Netflix
              </b>
              roulette
            </div>
          </div>
          <div>
            <button
              class="btn btn-primary"
              type="button"
            >
              + Add Movie
            </button>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="p-5 col"
          >
            <form
              class="d-flex justify-content-center"
            >
              <label
                for="movieName"
              >
                Find your movie
              </label>
              <input
                id="movieName"
                type="text"
                value=""
              />
              <button
                class="btn btn-primary"
                type="submit"
              >
                Search
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Header /> should Render header displaying a movie 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="d-flex"
          >
            <div
              class="me-auto p-2"
            >
              <div>
                <b>
                  Netflix
                </b>
                roulette
              </div>
            </div>
            <div>
              <button
                class="btn btn-primary"
                data-testid="search-movie-button"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="p-4 col-md-4"
            >
              <img
                class="card-img-top"
              />
            </div>
            <div
              class="p-5 title-movie-selected col"
            >
              <h1>
                A movie
                 - 
                <span
                  class="h4"
                />
              </h1>
              <h6 />
              <div
                class="p-3 row"
              >
                <div
                  class="col-2"
                />
                <div
                  class="col"
                />
              </div>
              <p>
                An overview
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="d-flex"
        >
          <div
            class="me-auto p-2"
          >
            <div>
              <b>
                Netflix
              </b>
              roulette
            </div>
          </div>
          <div>
            <button
              class="btn btn-primary"
              data-testid="search-movie-button"
              type="button"
            >
              <svg
                fill="currentColor"
                height="1em"
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="p-4 col-md-4"
          >
            <img
              class="card-img-top"
            />
          </div>
          <div
            class="p-5 title-movie-selected col"
          >
            <h1>
              A movie
               - 
              <span
                class="h4"
              />
            </h1>
            <h6 />
            <div
              class="p-3 row"
            >
              <div
                class="col-2"
              />
              <div
                class="col"
              />
            </div>
            <p>
              An overview
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Header /> should Render header displaying a movie and unselecting it when state change 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="d-flex"
          >
            <div
              class="me-auto p-2"
            >
              <div>
                <b>
                  Netflix
                </b>
                roulette
              </div>
            </div>
            <div>
              <button
                class="btn btn-primary"
                type="button"
              >
                + Add Movie
              </button>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="p-5 col"
            >
              <form
                class="d-flex justify-content-center"
              >
                <label
                  for="movieName"
                >
                  Find your movie
                </label>
                <input
                  id="movieName"
                  type="text"
                  value=""
                />
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  Search
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="d-flex"
        >
          <div
            class="me-auto p-2"
          >
            <div>
              <b>
                Netflix
              </b>
              roulette
            </div>
          </div>
          <div>
            <button
              class="btn btn-primary"
              type="button"
            >
              + Add Movie
            </button>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="p-5 col"
          >
            <form
              class="d-flex justify-content-center"
            >
              <label
                for="movieName"
              >
                Find your movie
              </label>
              <input
                id="movieName"
                type="text"
                value=""
              />
              <button
                class="btn btn-primary"
                type="submit"
              >
                Search
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
